### PR TITLE
Fix render order of components and add tests

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

[I am porting](https://github.com/bluefireteam/gravitational_waves/pull/4) Gravity Runner (aka Gravitational Waves) to the latest Flame.
I think it will be very good to have a fully functional, complete big game example before release.
Most of it was painless or even deleting stuff that we had incorporated from GR to Flame (like camera, viewport, etc).
However, I did come out with a list of things that were easier to do before. So I am putting out a few PRs.
And also I found this bug where component rendering order was not respected for isHud components.
This fixes that but in order to add a test I am modifying `canvas_test` to better test save/restore: https://github.com/bluefireteam/canvas_test/pull/3

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `flutter format` and the `flutter analyze` does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

**Review by spydon — APPROVED** [review]
Lgtm, what happens when a hud component is a child of a non-hud component?

**Comment by luanpotter** [comment]
@spydon that is a good question. I believe isHud is only considered for top level components. but that is existing behaviour, that we have to consider when we fully fix the camera. other point is that there is no way to bypass the viewport

**Comment by luanpotter** [comment]
(this requires the new version of canvas_test to pass)

File: packages/flame/test/game/component_rendering_test.dart [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import 'package:flame_test/flame_test.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class _MyComponent extends Component {
+  @override
+  bool isHud;
+
+  _MyComponent(int priority, {this.isHud = false}) : super(priority: priority);
```
Comment (erickzanardo) [inline packages/flame/test/game/component_rendering_test.dart:16]:
```suggestion
  _MyComponent(int priority, {this.isHud = false}) : super(priority: priority);
 
```

File: packages/flame/test/game/component_rendering_test.dart [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+          canvas,
+          MockCanvas()
+            ..translate(-12.0, -18.0)
+            ..drawRect(const Rect.fromLTWH(1, 1, 1, 1))
+            ..drawRect(const Rect.fromLTWH(2, 2, 1, 1))
+            ..drawRect(const Rect.fromLTWH(4, 4, 1, 1))
+            ..translate(0.0, 0.0),
+        );
+      },
+    );
```
Comment (erickzanardo) [inline packages/flame/test/game/component_rendering_test.dart:50]:
```suggestion
    );
  
```

File: packages/flame/test/game/component_rendering_test.dart [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        expect(
+          canvas,
+          MockCanvas()
+            ..drawRect(const Rect.fromLTWH(1, 1, 1, 1))
+            ..drawRect(const Rect.fromLTWH(2, 2, 1, 1))
+            ..drawRect(const Rect.fromLTWH(4, 4, 1, 1)),
+        );
+      },
+    );
```
Comment (erickzanardo) [inline packages/flame/test/game/component_rendering_test.dart:71]:
```suggestion
    );
  
```

**Review by erickzanardo — APPROVED** [review]

**Comment by st-pasha** [comment]
I am not sure what bug is being fixed here?

The `isHud` components are deliberately rendered after the non-hud components, so that they would visually appear over of the game world.

File: packages/flame/test/components/position_component_test.dart [hunk 4]
```diff
@@ -632,7 +632,8 @@ void main() {
           ..drawLine(const Offset(0, -2), const Offset(0, 2))
           ..drawLine(const Offset(-2, 0), const Offset(2, 0))
           ..drawParagraph(null, const Offset(-30, -15))
-          ..drawParagraph(null, const Offset(-20, 10)),
+          ..drawParagraph(null, const Offset(-20, 10))
+          ..translate(0, 0), // canvas.restore
```
Comment (st-pasha) [inline packages/flame/test/components/position_component_test.dart:636]: why is this needed?

**Comment by luanpotter** [comment]
@st-pasha  `isHud` just means do not respect camera. they should be rendered according to priority order, which was broken at some point, causing issues with hud components on Gravity.

File: packages/flame/test/components/position_component_test.dart (hunk 4 above)
Comment (luanpotter) [reply packages/flame/test/components/position_component_test.dart:636]: after the render cycle our implementation resets the camera, so this represents that change. we could ignore a final translate(0, 0) but I think it makes sense to add it to the tests as it does happen in reality

**Comment by st-pasha** [comment]
@luanpotter There was a bug not too long ago (#934) about the fact that `isHud` components must be rendered on top of the game world ignoring the priority, and they weren't. So this PR would re-introduce that bug.

I understand that there could be compelling reasons to render hud components inter-mixed with world components, but I cannot come up with such scenarios. Could you please clarify the use-case that you had in mind for this PR?

**Comment by spydon** [comment]
> @luanpotter There was a bug not too long ago (#934) about the fact that `isHud` components must be rendered on top of the game world ignoring the priority, and they weren't. So this PR would re-introduce that bug.
> 
> I understand that there could be compelling reasons to render hud components inter-mixed with world components, but I cannot come up with such scenarios. Could you please clarify the use-case that you had in mind for this PR?

It was my fault opening that bug, `isHud` should really be named `ignoreCamera` and later we should introduce `ignoreViewport` too. The parallax is an example for when it is needed.

**Comment by st-pasha** [comment]
> isHud should really be named ignoreCamera and later we should introduce ignoreViewport too.

"isHud" means the component is part of a "heads-up display", a UI that renders on top of the world. This is how it is currently defined, and used. Modifying that definition is a huge breaking change at this point.

Now, I agree with you that "isHud" is not a very good property. In particular, it only works for top-level components, which is a sure sign that it cannot be a generic property defined for every component.

However, I don't agree about renaming it, or introducing `ignoreViewport`: such properties would again suffer from the same problem that they won't work anywhere except at the top level of the game tree. Instead of encumbering `Component` class with more and more stuff, we should explore alternative solutions tied to the Camera / Viewport class.

> The parallax is an example for when it is needed.

`ParallaxComponent` is a very special component indeed. And it is long overdue for refactoring to make it more useful. It is definitely not a Hud component. However, it does interact with a camera in its own special way. It doesn't mean that the Camera should be aware of the ParallaxComponent though, instead the ParallaxComponent should be aware of the camera.

**Comment by luanpotter** [comment]
`isHUD` might be a bad naming, but the intent was always to just ignore the camera. That is what it should do. We already have priority to define priorities. It makes no sense to couple this concepts in an irreparable way, and it causes issues such as hacks in our camera system to account for parallax.

If you look at Gravity there are a handful of examples of HUD components that should definitely NOT be on top; thus upgrading flame broke all this use cases. If do you want your HUDs it to be on top just set the priority (we have that on Gravity too); the other way around, it's impossible to work around.

I agree with @spydon the name is not ideal and should be replaced with respectCamera and a respectViewport should be introduced, which is another issue that I pointed out (no way to do it right now).

But for now, I believe this PR as-is is a step forward as it's fixing a current bug.

We can consider the followups later, as I also agree with @st-pasha that the respect* properties are not perfect bc they are only relevant to top level components, but not sure what change we can easily do instead to properly fix this. Since we have a very tight timeline for v1 we might wanna go with that unless we can think of a better solution that is achievable rn.

**Comment by st-pasha** [comment]
> If you look at Gravity there are a handful of examples of HUD components that should definitely NOT be on top.

Could you please post some links to screenshots or videos of how this looks like? I'd like to understand the use case(s) here better.
